### PR TITLE
NIFI-4005 Add support for Azure Shared Access Signature (SAS) Tokens

### PR DESCRIPTION
**Pull request by aperepel**
- Refactoring and cleanup
- Upgraded the client library from 5.0.0 to 5.2.0

### CODE REVIEW

**Comment by jtstorck** [comment]
Reviewing...

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/Azure.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                    "one uses Shared Access Signature (SAS) token instead for fine-grained control with policies. " +
+                    "There are certain risks in allowing the account key to be stored as a flowfile" +
+                    "attribute. While it does provide for a more flexible flow by allowing the account key to " +
+                    "be fetched dynamically from a flow file attribute, care must be taken to restrict access to " +
+                    "the event provenance data (e.g. by strictly controlling the policies governing provenance for this Processor). " +
+                    "In addition, the provenance repositories may be put on encrypted disk partitions.")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR).expressionLanguageSupported(true).required(false).sensitive(true).build();
+
+    public static final PropertyDescriptor ACCOUNT_NAME = new PropertyDescriptor.Builder().name("storage-account-name").displayName("Storage Account Name")
+            .description("The storage account name.  There are certain risks in allowing the account name to be stored as a flowfile" +
```
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/Azure.java]: A whitespace is needed after the 'as a flowfile'. It's displayed as 'as a flowfileattribute.'

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/Azure.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import org.apache.nifi.processor.util.StandardValidators;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.InvalidKeyException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public final class Azure {
```
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/Azure.java]: The class name may be too broad IMHO. Probably `AzureStorageUtils` if it envisions to cover other storage services such as File, Queue or Table. Or `AzureBlobUtils` to be more specific. Thoughts?

File: nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/Azure.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                    "be fetched dynamically from a flow file attribute, care must be taken to restrict access to " +
+                    "the event provenance data (e.g. by strictly controlling the policies governing provenance for this Processor). " +
+                    "In addition, the provenance repositories may be put on encrypted disk partitions.")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR).expressionLanguageSupported(true).required(true).sensitive(true).build();
+
+    public static final PropertyDescriptor CONTAINER = new PropertyDescriptor.Builder().name("container-name").displayName("Container Name")
+            .description("Name of the Azure storage container").addValidator(StandardValidators.NON_EMPTY_VALIDATOR).expressionLanguageSupported(true).required(true).build();
+
+    public static final PropertyDescriptor PROP_SAS_TOKEN = new PropertyDescriptor.Builder()
+            .name("SAS String")
```
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/storage/utils/Azure.java]: Please use name and displayName as other properties do. "SAS Token" would be better go along with Azure terminology. Especially when I copied value from Azure Web console, I could be more confident where should I copy the value from if it's labeled as "SAS Token".
image

**Comment by ijokarumawak** [comment]
@aperepel I took over updating the PR. The new one #2353 is submitted based on your commit. Would you close this PR? Thanks!

**Comment by aperepel** [comment]
Thank you, Koji!

**Pull request closed without merging**